### PR TITLE
implement format on save

### PR DESCRIPTION
**Pull request by devoncarew**
- implement format on save
- also, implement organize imports on save
- fix https://github.com/flutter/flutter-intellij/issues/1205

screen shot 2018-03-02 at 3 12 12 pm

This might be a bit twitchy, so I added it in the experimental section.

@scheglov

### CODE REVIEW

**Comment by devoncarew** [comment]
This does send several calls to the analysis server, which we may find ways to optimize in the future (organize imports, format, then - because the we modified the file contents - another round of organize imports and format).

File: src/io/flutter/editor/FlutterSaveActionsManager.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+        if (reSave || didFormat) {
+          FileDocumentManager.getInstance().saveDocument(document);
+        }
+      }
+    }.execute());
+  }
+
+  private @Nullable
+  Module getModuleForPath(@NotNull VirtualFile file) {
```
Comment (scheglov) [inline src/io/flutter/editor/FlutterSaveActionsManager.java]: getModuleForFile?

File: src/io/flutter/editor/FlutterSaveActionsManager.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    if (PsiErrorElementUtil.hasErrors(myProject, psiFile.getVirtualFile())) {
+      return;
+    }
+
+    DartAnalysisServerService.getInstance(myProject).serverReadyForRequest(myProject);
+
+    if (settings.isOrganizeImportsOnSaveKey()) {
+      performOrganizeImportsOnSave(document, file);
+    }
+    else {
```
Comment (scheglov) [inline src/io/flutter/editor/FlutterSaveActionsManager.java]: So, we do either organize import, or format?
Maybe I don't understand IntelliJ UI format, but it says `JCheckBox`, so both can be checked?

Comment (scheglov) [reply src/io/flutter/editor/FlutterSaveActionsManager.java]: Ah, I see.
We do both when organize imports.

File: src/io/flutter/editor/FlutterSaveActionsManager.java (hunk 1 above)
Comment (alexander-doroshko) [reply src/io/flutter/editor/FlutterSaveActionsManager.java]: Go To Symbol -> `findModuleForFile` will show you an existing helper method that works better :)

Comment (devoncarew) [reply src/io/flutter/editor/FlutterSaveActionsManager.java]: Updated! Thanks :)

File: src/io/flutter/editor/FlutterSaveActionsManager.java (hunk 2 above)
Comment (devoncarew) [reply src/io/flutter/editor/FlutterSaveActionsManager.java]: I renamed the methods to make this clearer.

**Comment by devoncarew** [comment]
From testing:
- we do correctly ignore dart files that are not part of the project
- it is twitchy with very fast edits and saves (less than 100ms between the initial modification and the cms-s); not sure why - I tried adding some logic to use the document modification stamps, but this didn't improve the race condition

**Comment by filiph** [comment]
Where does one find this? I'm trying to search in IntelliJ setting but nothing finds this option.

**Comment by devoncarew** [comment]
It's in the Flutter plugin settings (`Languages & Frameworks` > `flutter`).

**Pull request merged**